From 4c6d617ba866f9801c735f7eff51d9941a9cc947 Mon Sep 17 00:00:00 2001
From: Fabio Estevam <fabio.estevam@freescale.com>
Date: Tue, 19 Jun 2012 20:17:26 -0300
Subject: [PATCH 32/32] ARM: imx23-olinuxino.dts: Allow configuring a gpio on
 boot time

GPIO0_17 must be set at boot time to level 1 so that LAN9215 can be functional.

Signed-off-by: Fabio Estevam <fabio.estevam@freescale.com>
---
 arch/arm/boot/dts/imx23-olinuxino.dts |    5 +++++
 arch/arm/boot/dts/imx23.dtsi          |    2 +-
 arch/arm/mach-mxs/mach-mxs.c          |   35 +++++++++++++++++++++++++++++++++
 3 files changed, 41 insertions(+), 1 deletion(-)

diff --git a/arch/arm/boot/dts/imx23-olinuxino.dts b/arch/arm/boot/dts/imx23-olinuxino.dts
index f7475bd..880129a 100644
--- a/arch/arm/boot/dts/imx23-olinuxino.dts
+++ b/arch/arm/boot/dts/imx23-olinuxino.dts
@@ -18,6 +18,11 @@
 	model = "i.MX23 Olinuxino Low Cost Board";
 	compatible = "olimex,imx23-olinuxino", "fsl,imx23";
 
+	config-on-boot {
+		output-gpios = <&gpio0 17 0>;	/* vbus power */
+		output-gpio-values = <1>;
+	};
+
 	memory {
 		reg = <0x40000000 0x04000000>;
 	};
diff --git a/arch/arm/boot/dts/imx23.dtsi b/arch/arm/boot/dts/imx23.dtsi
index c4f7c88..62df88c 100644
--- a/arch/arm/boot/dts/imx23.dtsi
+++ b/arch/arm/boot/dts/imx23.dtsi
@@ -114,7 +114,7 @@
 
 				duart_pins_a: duart@0 {
 					reg = <0>;
-					fsl,pinmux-ids = <0x11a2 0x11b2>;
+					fsl,pinmux-ids = <0x11a2 0x11b2 0x0113>;
 					fsl,drive-strength = <0>;
 					fsl,voltage = <1>;
 					fsl,pull-up = <0>;
diff --git a/arch/arm/mach-mxs/mach-mxs.c b/arch/arm/mach-mxs/mach-mxs.c
index dc276bb..8d5e5fe 100644
--- a/arch/arm/mach-mxs/mach-mxs.c
+++ b/arch/arm/mach-mxs/mach-mxs.c
@@ -18,6 +18,7 @@
 #include <linux/irqdomain.h>
 #include <linux/of_irq.h>
 #include <linux/of_platform.h>
+#include <linux/of_gpio.h>
 #include <asm/mach/arch.h>
 #include <asm/mach/time.h>
 #include <mach/common.h>
@@ -71,6 +72,38 @@ static struct sys_timer imx28_timer = {
 	.init = imx28_timer_init,
 };
 
+static void __init mxs_config_on_boot(void)
+{
+	struct device_node *np;
+	struct property *pp;
+	int cnt, len, i;
+	int gpio;
+
+	np = of_find_node_by_path("/config-on-boot");
+	if (!np)
+		return;
+	cnt = of_gpio_named_count(np, "output-gpios");
+	pp = of_find_property(np, "output-gpio-values", &len);
+	if (!pp || cnt != len / sizeof(u32)) {
+		pr_err("Invalid config-on-boot gpios!\n");
+		of_node_put(np);
+		return;
+	}
+	for (i = 0; i < cnt; i++) {
+		gpio = of_get_named_gpio(np, "output-gpios", i);
+		if (gpio_is_valid(gpio))
+			gpio_request_one(gpio, GPIOF_OUT_INIT_HIGH,
+					"config-on-boot");
+	}
+
+	of_node_put(np);
+}
+
+static void __init mxs_post_populate(void)
+{
+	mxs_config_on_boot();
+}
+
 static void __init imx28_evk_init(void)
 {
 	struct clk *clk;
@@ -88,6 +121,8 @@ static void __init mxs_machine_init(void)
 
 	of_platform_populate(NULL, of_default_bus_match_table,
 				NULL, NULL);
+
+	mxs_post_populate();
 }
 
 static const char *imx23_dt_compat[] __initdata = {
-- 
1.7.10

